events.js

```diff
@@ -1,472 +1,488 @@
+      if(rep && rep.tx && rep.sent == lseq - 1 && lseq > seq) {
+        peer.msgs.push(msg)
+        rep.sent++
+      }
+      //if we are ahead of this peer, and not in tx mode, let them know that.
+      else if(
+        isAhead(lseq, seq) &&
+          (rep ? !rep.tx && rep.sent != null : state.follows[author])
+      )
+        setNotes(peer, author, sequence, false)
     }
-      //XXX activate some other peer?
+
     return state
   }
 
-  //remember the time of the last message received
-  state.peers[ev.id].ts = ev.ts
+  //XXX if we only receive from a single peer,
+  //then we shouldn't really get known messages?
+  //except during the race when we have disabled a peer
+  //but they havn't noticed yet.
+  exports.receive = function (state, ev) {
+    var msg = ev.value
+    //receive a message, validate and append.
+    //if this message is forked, disable this feed
+
+    if(!state.peers[ev.id]) throw new Error('lost peer state:'+ev.id)
+
+    //we _know_ that this peer is upto at least this message now.
+    //(but maybe they already told us they where ahead further)
+    const author = exports.getMsgAuthor(msg)
+    const sequence = exports.getMsgSequence(msg)
+    var peer = state.peers[ev.id]
+    var rep = peer.replicating[author]
+
+    //if we havn't asked for this, ignore it. (this is remote speaking protocol wrong!)
+    if(!rep) return state
+
+    peer.clock[author] = Math.max(peer.clock[author], sequence)
+    rep.sent = Math.max(rep.sent, sequence)
+
+    //if this message has already been seen, ignore.
+    if(state.clock[author] >= sequence) {
+      if (rep.rx) {
+        setNotes(peer, author, state.clock[author], false)
+      }
+      //XXX activate some other peer?
+      return state
+    }
+
+    //remember the time of the last message received
+    state.peers[ev.id].ts = ev.ts
 
-  //FORKS ignore additional messages if we have already found an invalid one.
-  if(isShared(state, ev.value.author, ev.id))
-    state.receive.push(ev)
-  //Q: possibly update the receiving mode?
+    //FORKS ignore additional messages if we have already found an invalid one.
+    if(isShared(state, exports.getMsgAuthor(ev.value), ev.id))
+      state.receive.push(ev)
+    //Q: possibly update the receiving mode?
 
-  return state
-}
+    return state
+  }
 
 
-//XXX check if we are already receiving a feed
-//and if so put this into lazy mode.
-exports.notes = function (state, ev) {
-  //update replicating modes
-  var clock = ev.value
+  //XXX check if we are already receiving a feed
+  //and if so put this into lazy mode.
+  exports.notes = function (state, ev) {
+    //update replicating modes
+    var clock = ev.value
 
-  //support sending clocks inside a thing with additional properties.
-  //this is to allow room for backwards compatible upgrades.
-  if(isObject(ev.value.clock))
-    clock = ev.value.clock
+    //support sending clocks inside a thing with additional properties.
```

Comment:
this is about the protocol, right? IMHO deserves a place in the README

Reply:
Agree. I've created an issue for that. This was also already there.